File: src/threads/threads.ts
import { cpus } from "os";
import { Worker, isMainThread, parentPort, workerData } from "worker_threads";

type Range = { start: number; end: number };

const splitWorkload = (total: number, workers: number): Range[] => {
  if (total <= 0) return [];
  const chunkSize = Math.ceil(total / workers);
  const ranges: Range[] = [];
  for (let start = 0; start < total; start += chunkSize) {
    ranges.push({ start, end: Math.min(start + chunkSize, total) });
  }
  return ranges;
};

const runWorkers = (ranges: Range[]): Promise<number[]> => {
  return Promise.all(
    ranges.map(
      (range) =>
        new Promise<number>((resolve, reject) => {
          const worker = new Worker(__filename, { workerData: range });
          worker.once("message", (value: number) => resolve(value));
          worker.once("error", reject);
        })
    )
  );
};

const main = async (): Promise<void> => {
  const totalNumbers = 100_000_000;
  const cores = cpus().length || 1;
  const workers = cores;

  const ranges = splitWorkload(totalNumbers, workers);
  if (ranges.length === 0) {
    console.log("Nothing to sum.");
    return;
  }

  const partials = await runWorkers(ranges);
  const total = partials.reduce((acc, value) => acc + value, 0);

  console.log(`Summed numbers 0..${totalNumbers - 1}`);
  console.log(`Workers used: ${ranges.length} (detected cores: ${cores})`);
  console.log(`Result: ${total}`);
};

const workerTask = (): void => {
  const { start, end } = workerData as Range;
  let subtotal = 0;
  for (let i = start; i < end; i++) {
    subtotal += i;
  }
  parentPort?.postMessage(subtotal);
};

if (isMainThread) {
  main().catch((error) => {
    console.error(error);
    process.exit(1);
  });
} else {
  workerTask();
}

File: src/threads/threads.py
import argparse
import concurrent.futures
import math
import os
from typing import Iterable, List, Optional, Tuple


def split_workload(total: int, workers: int) -> List[Tuple[int, int]]:
    """Split a range [0, total) into nearly even chunks."""
    if total <= 0:
        return []

    workers = max(1, workers)
    chunk_size = math.ceil(total / workers)
    return [
        (start, min(start + chunk_size, total))
        for start in range(0, total, chunk_size)
    ]


def partial_sum(bounds: Tuple[int, int]) -> int:
    start, end = bounds
    result = sum(range(start, end))
    print(f'start={start} | end={end} | result={result}\n')
    return result


def sum_numbers(total: int, workers: Optional[int] = None) -> int:
    if workers is None:
        workers = os.cpu_count() or 1

    ranges = split_workload(total, workers)
    print(ranges)
    if not ranges:
        return 0

    # Threads: good for IO; CPU-bound work may still be limited by the GIL.
    with concurrent.futures.ThreadPoolExecutor(max_workers=len(ranges)) as pool:
        partials: Iterable[int] = pool.map(partial_sum, ranges)
    return sum(partials)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Sum numbers using threads.")
    parser.add_argument(
        "--total",
        type=int,
        default=100_000_000,
        help="Total count of sequential numbers to sum (default: 100,000,000).",
    )
    parser.add_argument(
        "--workers",
        type=int,
        default=None,
        help="Number of worker threads (default: CPU count).",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    cores = os.cpu_count() or 1
    workers = args.workers or cores
    total = args.total

    result = sum_numbers(total, workers)

    print(f"Summed numbers 0..{args.total - 1}")
    print(f"Workers used: {workers} (detected cores: {cores})")
    print(f"Result: {result}")


if __name__ == "__main__":
    main()
